src/main/scala/com/microsoft/hyperspace/HyperspaceSparkSessionExtension.scala

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (2020) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace
+
+import org.apache.spark.sql.{SparkSession, SparkSessionExtensions}
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * An extension for Spark SQL to activate Hyperspace.
+ *
+ * Example to run a `spark-submit` with Hyperspace enabled:
+ * {{{
+ *   spark-submit -c spark.sql.extensions=com.microsoft.hyperspace.HyperspaceSparkSessionExtension
+ * }}}
+ *
+ * Example to create a `SparkSession` with Hyperspace enabled:
+ * {{{
+ *    val spark = SparkSession
+ *       .builder()
+ *       .appName("...")
+ *       .master("...")
+ *       .config("spark.sql.extensions", "com.microsoft.hyperspace.HyperspaceSparkSessionExtension")
+ *       .getOrCreate()
+ * }}}
+ */
+
+class HyperspaceSparkSessionExtension extends (SparkSessionExtensions => Unit) {
+
+  /**
+   * If HyperspaceRule is injected directly to OptimizerRule with HyperspaceExtension,
+   * the order of applying rule is different from without HyperspaceExtension
+   * (i.e., explicitly calling enableHyperspace). To make behavior consistently,
+   * current implementation of HyperspaceExtension uses a trick to call enableHyperspace
+   * before rule is applied. Since the interface of injectOptimizerRule should return rule builder,
+   * it returns a dummy rule that does nothing. It may increase overhead slightly
+   * because enableHyperspace is called once for each evaluation of spark plan.
+   */
+  private case class DummyRule() extends Rule[LogicalPlan] {
+    override def apply(plan: LogicalPlan): LogicalPlan = {
+      plan
+    }
+  }
+
+  type RuleBuilder = SparkSession => Rule[LogicalPlan]
+  override def apply(extensions: SparkSessionExtensions): Unit = {
+    extensions.injectOptimizerRule { sparkSession =>
+      // Enable Hyperspace to leverage indexes.
+      sparkSession.enableHyperspace()
```

Comment:
According to the new model, `enableHyperspace` only exists for backward compatibility. I think it's better to factor out the rule insertion code out of this method and invoke the method here and from `enableHyperspace`.

Reply:
Do you mean this part?
```
      if (!sparkSession.sessionState.experimentalMethods.extraOptimizations.contains(
          ApplyHyperspace)) {
        sparkSession.sessionState.experimentalMethods.extraOptimizations ++=
          ApplyHyperspace :: Nil
      }
      if (!sparkSession.sessionState.experimentalMethods.extraStrategies.contains(
          BucketUnionStrategy)) {
        sparkSession.sessionState.experimentalMethods.extraStrategies ++=
          BucketUnionStrategy :: Nil
      }
```
Where should I put this code because `package object hyperspace` is quite customer side interfaces, so not sure if it is ok to create a function like

```
package object hyperspace {

  /**
   * Hyperspace-specific implicit class on SparkSession.
   */
  implicit class Implicits(sparkSession: SparkSession) {

    def enableHyperspace(): SparkSession = {
      HyperspaceConf.setHyperspaceApplyEnabled(sparkSession, true)
      addOptimizationsIfNeeded()
      sparkSession
    }

    private def addOptimizationsIfNeeded(): Unit = {
      if (!sparkSession.sessionState.experimentalMethods.extraOptimizations.contains(
          ApplyHyperspace)) {
        sparkSession.sessionState.experimentalMethods.extraOptimizations ++=
          ApplyHyperspace :: Nil
      }
      if (!sparkSession.sessionState.experimentalMethods.extraStrategies.contains(
          BucketUnionStrategy)) {
        sparkSession.sessionState.experimentalMethods.extraStrategies ++=
          BucketUnionStrategy :: Nil
      }
    }
  }
}
```

Reply:
You can put `addOptimizationsIfNeeded()` in a companion object `HyperspaceSparkSessionExtension` and call the method from here and `enableHyperspace()`.

Reply:
@clee704 I've update the functions so please check whether the implement was done as you expected. Thanks!

Reply:
@clee704  can you check whether I implemented as you expected?